### PR TITLE
Optional better UI scaling

### PR DESCRIPTION
**Pull request by axbmcuser**
Ref.: https://github.com/xbmc/xbmc/pull/18780

Hi fellow Kodi supporters, 

i hope you are all well. After another 4 years flying by and still repeatedly being asked by different ppl and project contributors that i please should try again to get the UI scaling improvements updated and merged, this PR is now a follow-up on https://github.com/xbmc/xbmc/pull/18780

Seeing the recently introduced new gui-advancedsettings.xml entries, i simplified and updated the implementation for latest git master, minimized the number of settings after further evaluation and moved them to the gui-section. 

While i fully understand the teased theoretical reservations about memory and performance of mipmaps in general, i can say that i already extensively tested this implementation since 8 years ago on as low as one of my very old, outdated low powered 2-core Intel NUC with merely 2GB of RAM with great results while using edge-case test-libraries with 2500+ entries of lanczos-scaled, full quality imageres-1080 posters/artwork - even in the initial phase when starting with an empty userdata/Thumbnails folder - and all this while using wall views with high amounts of items displayed at once. 

To lower any hurdles, i left the implementation _fully_ _optional_ by default. 

Bottom line key facts:

- It's only a few lines of code in very few files without any complex changes to the source

- It's extremely easily reversible _if_ one day someone writes a completely different and new implementation
(which even seems unlikely for now considering that 16(!) years have gone by since discussions about ui scaling started)

- Potential risks: Near zero to none

- Proven benefit for a cleanly scaled UI: Huge

&nbsp;
Cheers

### CODE REVIEW

**Comment by sarbes** [comment]
I don't think that mipmapping is the right approach. Especially since this happens in a time critical path, which will stall the rendering on low-end systems. Actually, I want to get rid of all mipmapping.

The right approach in my eyes is to scale images to the resolution they are used at. This way, you get less jank, better performance, lower memory usage and (in most cases) better image quality. It is applicable to all rendering platforms. And this doesn't require a "completely different and new implementation", but a modification of our current scaling system.

**Comment by zag2me** [comment]
Greatly appreciate the GUI setting finally. This really does improve the overall look of Kodi.

**Comment by jjd-uk** [comment]
I assume the RPi is still our reference low end system? As this is optional and not enabled by default I assume that addresses @popcornmix concerns from original PR https://github.com/xbmc/xbmc/pull/8409#issuecomment-158380637 however those views were from a time when the RPi2 was the latest at that time.

**Comment by sarbes** [comment]
Kodi has already the capability to scale textures. You can see it in action when an uncached texture gets displayed. It scales to the appropriate resolution.

The path of loading a cached texture is lacking this feature. Cached textures get uploaded at the resolution they are stored at. They should receive similar treatment instead.

I'm sorry, but I don't see why we should have this workaround instead of fixing the issue at its root. A fix should be of similar complexity.

**Comment by sarbes** [comment]
> I assume the RPi is still our reference low end system? As this is optional and not enabled by default I assume that addresses @popcornmix concerns from original PR [#8409 (comment)](https://github.com/xbmc/xbmc/pull/8409#issuecomment-158380637) however those views were from a time when the RPi2 was the latest at that time.

I would assume that even on more capable SOCs, this would introduce jank. It heavily depends on texture resolution, driver implementation and system load. But stalling the CPU or the GPU to generate mips is never a good idea.

On a lot of GPUs, it would double the shader execution time, potentially halving the frame rate.

But in the end, this PR doesn't implement it on GLES, so the question is purely theoretical at this point.

**Comment by popcornmix** [comment]
Yes, as this specific PR doesn't affect GLES then it won't affect behaviour on any Pi/arm devices, so no objection there.

I do agree with @sarbes that having an oversized texture uploaded to the GPU, then having to generate mipmaps to allow for the  actually desired smaller sized texture to be displayed without aliasing is a very inefficient way to go.

Just uploading a texture of the desired size would be much better (especially if the resized texture goes via the texture cache, so is a one-off cost) in both memory usage and fps.

But it depends on whether there is a plan to implement the "correct" solution.

**Comment by sarbes** [comment]
> Just uploading a texture of the desired size would be much better (especially if the resized texture goes via the texture cache, so is a one-off cost) in both memory usage and fps.

The term "texture cache" is a bit misleading, to be honest. We cache image assets in their original form, or at reduced resolution/quality if they exceed certain limits. Image cache would be more appropriate.

We do derive our textures from this source at runtime. But this is totally fine, as this is done in its own thread at low priority. 


Also, mipmapping has interactions with other rendering features. This is not something I want to deal with when working on our GUI system.

**Comment by neo1973** [comment]
This has become obsolete with the merge of #25630.
